### PR TITLE
Kubevirt Console Plugin Integration with HCO

### PR DESCRIPTION
**Pull request by orenc1**
Add Operand Handlers to create and reconcile new resources for the Kubvirt console plugin in OpenShift:
1. Deployment for the kubevirt plugin, serving the UI contents using nginx
2. Service that exposes the kubevirt plugin pod
3. Custom Resource for the plugin - `ConsolePlugin` (`console.openshift.io/v1alpha1`)

**Note:**
kubevirt console plugin resources will be deployed only if `KV_CONSOLE_PLUGIN_IMAGE` environment variable is provided to hco-operator, with the value of the plugin image pullspec. Currently it is not activated, until an official plugin release will be available at https://github.com/kubevirt-ui/kubevirt-plugin/releases.

Signed-off-by: orenc1 <ocohen@redhat.com>

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly
  
**Release note**:

```release-note
kubevirt console plugin integration with HCO
```

### CODE REVIEW

File: hack/config [hunk 1]
```diff
@@ -13,6 +13,7 @@ LIVENESS_PROBE_VERSION="v2.3.0"
 CSI_SNAPSHOT_VERSION="v4.2.1"
 CSI_SIG_STORAGE_PROVISIONER_VERSION="v2.2.1"
 KUBE_RBAC_PROXY_VERSION="4.10"
+KUBEVIRT_PLUGIN_VERSION="latest"
```
Comment (yaacov) [inline hack/config:16]: maybe we should add a "UI" or "CONSOLE" to the name, as it is it can get confused with the main kubevirt operators,
for example ''KUBEVIRT_CONSOLE_PLUGIN_VERSION", does that make sense to you ?

File: controllers/operands/kubevirtConsolePlugin.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+		if err != nil {
+			return false, false, err
+		}
+		updated = true
+	}
+
+	// Enable console plugin for kubevirt if not already enabled
+	consoleKey := client.ObjectKey{Namespace: hcoutil.UndefinedNamespace, Name: "cluster"}
+	consoleObj := &operatorv1.Console{}
+	err := Client.Get(req.Ctx, consoleKey, consoleObj)
```
Comment (orenc1) [inline controllers/operands/kubevirtConsolePlugin.go:197]: When it is being run in the context of the unit-test, the `consoles.operator.openshift.io` resource could not be fetched, and I'm getting:
```
"consoles.operator.openshift.io \"cluster\" not found"
```
Although I'm adding the resource in:
https://github.com/kubevirt/hyperconverged-cluster-operator/pull/1809/files#diff-7e657fcfc15ca5a4e8518504c3b763bf67b211a6a3577c59881b37e557377c96R195-R197

```
	expectedConsoleConfig := &operatorv1.Console{
		TypeMeta: metav1.TypeMeta{
			Kind:       "Console",
			APIVersion: "operator.openshift.io/v1",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name: "cluster",
		},
	}
	res.consoleConfig = expectedConsoleConfig
```

Suggestions?

**Review by nunnatsa — COMMENTED** [review]
Added some inline comments.

File: pkg/util/consts.go [hunk 3]
```diff
@@ -59,6 +59,7 @@ const (
 	DefaultWebhookCertDir = "/apiserver.local.config/certificates"
 
 	CliDownloadsServerPort = 8080
+	UiPluginServerPort     = 8082
```
[the remark this answers is not in the document]
Comment (orenc1) [reply pkg/util/consts.go:62]: Also do the same for all other port constants (e.g. `WebhookPort`, `CliDownloadsServerPort`) ?

File: controllers/operands/kubevirtConsolePlugin.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	err := h.Client.Get(req.Ctx, consoleKey, consoleObj)
+	if err != nil {
+		req.Logger.Error(err, fmt.Sprintf("Could not find resource - APIVersion: %s, Kind: %s, Name: %s",
+			consoleObj.APIVersion, consoleObj.Kind, consoleObj.Name))
+		return &EnsureResult{
+			Err: nil,
+		}
+	}
+
+	plugins := consoleObj.Spec.Plugins
```
[the remark this answers is not in the document]
Comment (orenc1) [reply controllers/operands/kubevirtConsolePlugin.go:225]: It was initially defined to shorten the reference to the plugins array.
But you're right, it doesn't really needed. Removed.

File: controllers/operands/kubevirtConsolePlugin.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+					Labels: map[string]string{
+						"app": kvUIPluginName,
+					},
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: "default",
+					Containers: []corev1.Container{
+						{
+							Name:            kvUIPluginName,
+							Image:           "quay.io/kubevirt-ui/kubevirt-plugin:latest",
```
[the remark this answers is not in the document]
Comment (orenc1) [reply controllers/operands/kubevirtConsolePlugin.go:85]: you mean in `csv-merger`.
done, thanks :+1:

File: controllers/operands/kubevirtConsolePlugin.go [hunk 6]
```diff
@@ -251,12 +249,11 @@ func (h consoleHandler) ensure(req *common.HcoRequest) *EnsureResult {
...
+func newConsoleHandler(Client client.Client, _ *runtime.Scheme) Operand {
 	h := &consoleHandler{
 		Client: Client,
-		Scheme: Scheme,
 	}
-	return []Operand{h}, nil
+	return h
 }
 
 func int32Ptr(i int32) *int32 {
```
Comment (nunnatsa) [inline controllers/operands/kubevirtConsolePlugin.go:259]: use:
```golang
import "k8s.io/utils/pointer"
...
pointer.Int32Ptr()
```
instead.

File: controllers/operands/kubevirtConsolePlugin.go [hunk 7]
```diff
@@ -55,6 +52,10 @@ func newKvUiPluginCRHandler(_ log.Logger, Client client.Client, Scheme *runtime.
 }
 
 func NewKvUiPluginDeplymnt(hc *hcov1beta1.HyperConverged) (*appsv1.Deployment, error) {
+	kvUiPluginImage, varExists := os.LookupEnv(hcoutil.KvUiPluginImageEnvV)
+	if !varExists {
+		return nil, errors.New(fmt.Sprintf("%s env var not found", hcoutil.KvUiPluginImageEnvV))
```
Comment (nunnatsa) [inline controllers/operands/kubevirtConsolePlugin.go:57]: This is a problem because in this case, HCO is in error loop that can't be resolved.

I think there are three options here:
1. Check it on boot and exit with error. That way a wrong image will fail in any basic test
2. only log here, but don't return error.
3. some combination of the first two

Comment (orenc1) [reply controllers/operands/kubevirtConsolePlugin.go:57]: implemented both approaches.

File: controllers/hyperconverged/hyperconverged_controller.go [hunk 8]
```diff
@@ -106,7 +105,7 @@ func newReconciler(mgr manager.Manager, ci hcoutil.ClusterInfo, upgradeableCond
 	return &ReconcileHyperConverged{
 		client:               mgr.GetClient(),
 		scheme:               mgr.GetScheme(),
-		operandHandler:       operands.NewOperandHandler(mgr.GetClient(), mgr.GetScheme(), ci.IsOpenshift(), hcoutil.GetEventEmitter()),
+		operandHandler:       operands.NewOperandHandler(mgr.GetClient(), mgr.GetScheme(), ci.IsOpenshift(), ci.IsConsolePluginImageProvided(), hcoutil.GetEventEmitter()),
```
Comment (nunnatsa) [inline controllers/hyperconverged/hyperconverged_controller.go:108]: What do you think about sending the whole `ci` object to the function? now there are 5 parameters already, some of them with a long name.

File: controllers/hyperconverged/hyperconverged_controller.go [hunk 9]
```diff
@@ -972,7 +971,8 @@ func getNumOfChangesJSONPatch(jsonPatch string) int {
 
 func (r *ReconcileHyperConverged) firstLoopInitialization(request *common.HcoRequest) {
 	// Initialize operand handler.
-	r.operandHandler.FirstUseInitiation(r.scheme, hcoutil.GetClusterInfo().IsOpenshift(), request.Instance)
+	r.operandHandler.FirstUseInitiation(r.scheme, hcoutil.GetClusterInfo().IsOpenshift(),
```
Comment (nunnatsa) [inline controllers/hyperconverged/hyperconverged_controller.go:974]: same

File: pkg/util/cluster.go [hunk 10]
```diff
@@ -59,6 +61,13 @@ func (c *ClusterInfoImp) Init(ctx context.Context, cl client.Client, logger logr
 		err = c.initKubernetes(cl)
 	}
 
+	varValue, varExists := os.LookupEnv(KvUiPluginImageEnvV)
+	if !varExists || varExists && len(varValue) > 0 {
+		c.consolePluginImageProvided = true
+	} else {
+		c.consolePluginImageProvided = false
+	}
```
Comment (nunnatsa) [inline pkg/util/cluster.go:69]:
```suggestion
	c.consolePluginImageProvided = !varExists || varExists && len(varValue) > 0
```

File: controllers/hyperconverged/hyperconverged_controller.go (hunk 8 above)
Comment (orenc1) [reply controllers/hyperconverged/hyperconverged_controller.go:108]: Done.

File: pkg/util/cluster.go (hunk 10 above)
Comment (orenc1) [reply pkg/util/cluster.go]: actually the logic wasn't quite accurate.
we want to set this flag to `true` if and only if both the var exists and its value has a non-zero length.
thanks

File: controllers/operands/operandHandler.go [hunk 11]
```diff
@@ -67,6 +66,12 @@ func NewOperandHandler(client client.Client, scheme *runtime.Scheme, isOpenshift
 		}...)
 	}
 
+	if ci.IsOpenshift() && ci.IsConsolePluginImageProvided() {
+		operands = append(operands, []Operand{
+			newConsoleHandler(client),
+		}...)
+	}
+
```
Comment (nunnatsa) [inline controllers/operands/operandHandler.go:74]: Not important: can be embedded within the previous condition. e.g. 
```suggestion
	if ci.IsOpenshift() {
		operands = append(operands, []Operand{
			newSspHandler(client, scheme),
			(*genericOperand)(newMetricsServiceHandler(client, scheme)),
			(*genericOperand)(newMetricsServiceMonitorHandler(client, scheme)),
			(*genericOperand)(newMonitoringPrometheusRuleHandler(client, scheme)),
			(*genericOperand)(newCliDownloadHandler(client, scheme)),
			(*genericOperand)(newCliDownloadsRouteHandler(client, scheme)),
			(*genericOperand)(newCliDownloadsServiceHandler(client, scheme)),
			newNamespaceHandler(client, scheme),
		}...)

	        if ci.IsConsolePluginImageProvided() {
		    operands = append(operands, []Operand{
			    newConsoleHandler(client),
		    }...)
		}
	}

```

But we can do it in a separate PR.

File: controllers/operands/operandHandler.go [hunk 12]
```diff
@@ -67,6 +66,12 @@ func NewOperandHandler(client client.Client, scheme *runtime.Scheme, isOpenshift
 		}...)
 	}
 
+	if ci.IsOpenshift() && ci.IsConsolePluginImageProvided() {
+		operands = append(operands, []Operand{
+			newConsoleHandler(client),
+		}...)
```
Comment (nunnatsa) [inline controllers/operands/operandHandler.go:72]: the 2nd parameter does not have to be a slice.
```suggestion
		operands = append(operands, newConsoleHandler(client))
```

**Review by nunnatsa — APPROVED** [review]
/approve

**Comment by orenc1** [comment]
/hold
just figured out the plugin service **must** use HTTPS to serve the plugin contents:
https://github.com/openshift/enhancements/blob/master/enhancements/console/dynamic-plugins.md#delivering-plugins
Will update accordingly.

**Comment by orenc1** [comment]
HCO is now configuring the plugin deployment and service to support HTTPS communications.
Additional handler has been created for the config map holding the nginx configurations.
/unhold
